File: lib/core/models/promo_models.dart
class PromoCodeConfig {
  const PromoCodeConfig({
    required this.code,
    required this.enabled,
    required this.autoAssignOnSignup,
    required this.discountPercent,
    required this.maxDiscountIqd,
    required this.maxRides,
    this.description = '',
    this.expiresAt,
    this.maxTotalRedemptions,
    this.currentRedemptions = 0,
    this.districtIds = const [],
    this.minCompletedRidesForEligibility = 0,
    this.kind = 'both',
  });

  final String code;
  final bool enabled;
  final bool autoAssignOnSignup;
  final int discountPercent;
  final int maxDiscountIqd;
  final int maxRides;
  final String description;
  final DateTime? expiresAt;
  final int? maxTotalRedemptions;
  final int currentRedemptions;
  final List<String> districtIds;
  final int minCompletedRidesForEligibility;
  /// `ride`, `delivery`, or `both`
  final String kind;

  bool get isExpired =>
      expiresAt != null && DateTime.now().isAfter(expiresAt!);

  bool get isRedemptionLimitReached {
    if (maxTotalRedemptions == null) return false;
    return currentRedemptions >= maxTotalRedemptions!;
  }

  static const free3Defaults = PromoCodeConfig(
    code: 'FREE3',
    enabled: true,
    autoAssignOnSignup: true,
    discountPercent: 50,
    maxDiscountIqd: 1000,
    maxRides: 2,
    description: '50% off first 2 rides (max 1,000 IQD each)',
    kind: 'ride',
  );

  factory PromoCodeConfig.fromMap(Map<String, dynamic>? data) {
    if (data == null) return free3Defaults;
    return PromoCodeConfig(
      code: data['code'] as String? ?? free3Defaults.code,
      enabled: data['enabled'] as bool? ?? true,
      autoAssignOnSignup: data['autoAssignOnSignup'] as bool? ?? true,
      discountPercent: (data['discountPercent'] as num?)?.toInt() ??
          free3Defaults.discountPercent,
      maxDiscountIqd: (data['maxDiscountIqd'] as num?)?.toInt() ??
          free3Defaults.maxDiscountIqd,
      maxRides: (data['maxRides'] as num?)?.toInt() ?? free3Defaults.maxRides,
      description: data['description'] as String? ?? '',
      expiresAt: (data['expiresAt'] as dynamic)?.toDate() as DateTime?,
      maxTotalRedemptions: (data['maxTotalRedemptions'] as num?)?.toInt(),
      currentRedemptions: (data['currentRedemptions'] as num?)?.toInt() ?? 0,
      districtIds: (data['districtIds'] as List<dynamic>?)
              ?.map((value) => value.toString())
              .toList() ??
          const [],
      minCompletedRidesForEligibility:
          (data['minCompletedRidesForEligibility'] as num?)?.toInt() ?? 0,
      kind: data['kind'] as String? ?? 'both',
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'code': code,
      'enabled': enabled,
      'autoAssignOnSignup': autoAssignOnSignup,
      'discountPercent': discountPercent,
      'maxDiscountIqd': maxDiscountIqd,
      'maxRides': maxRides,
      'description': description,
      if (expiresAt != null) 'expiresAt': expiresAt,
      if (maxTotalRedemptions != null)
        'maxTotalRedemptions': maxTotalRedemptions,
      'currentRedemptions': currentRedemptions,
      'districtIds': districtIds,
      'minCompletedRidesForEligibility': minCompletedRidesForEligibility,
      'kind': kind,
    };
  }
}

class PromoApplication {
  const PromoApplication({
    required this.baseFareIqd,
    required this.discountIqd,
    required this.finalFareIqd,
    this.promoCode = '',
  });

  final int baseFareIqd;
  final int discountIqd;
  final int finalFareIqd;
  final String promoCode;

  bool get hasDiscount => discountIqd > 0 && promoCode.isNotEmpty;
}

class MonthlyPrizeConfig {
  const MonthlyPrizeConfig({
    required this.prizeAmountIqd,
    required this.monthKey,
    this.winnerDriverId = '',
    this.winnerPaid = false,
  });

  final int prizeAmountIqd;
  final String monthKey;
  final String winnerDriverId;
  final bool winnerPaid;

  static const defaultPrizeIqd = 50000;

  factory MonthlyPrizeConfig.fromMap(Map<String, dynamic>? data) {
    if (data == null) {
      return MonthlyPrizeConfig(
        prizeAmountIqd: defaultPrizeIqd,
        monthKey: MonthlyPrizeConfig.currentMonthKey(),
      );
    }
    return MonthlyPrizeConfig(
      prizeAmountIqd:
          (data['prizeAmountIqd'] as num?)?.toInt() ?? defaultPrizeIqd,
      monthKey: data['monthKey'] as String? ?? currentMonthKey(),
      winnerDriverId: data['winnerDriverId'] as String? ?? '',
      winnerPaid: data['winnerPaid'] as bool? ?? false,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'prizeAmountIqd': prizeAmountIqd,
      'monthKey': monthKey,
      'winnerDriverId': winnerDriverId,
      'winnerPaid': winnerPaid,
    };
  }

  static String currentMonthKey([DateTime? date]) {
    final value = date ?? DateTime.now();
    final month = value.month.toString().padLeft(2, '0');
    return '${value.year}-$month';
  }
}

class MonthlyLeaderboardEntry {
  const MonthlyLeaderboardEntry({
    required this.driverId,
    required this.name,
    required this.phone,
    required this.rideCount,
    required this.rank,
    required this.isWinner,
    required this.isPaid,
  });

  final String driverId;
  final String name;
  final String phone;
  final int rideCount;
  final int rank;
  final bool isWinner;
  final bool isPaid;
}

class LoyaltyConfig {
  const LoyaltyConfig({
    this.enabled = false,
    this.ridesRequired = 10,
    this.repeats = true,
  });

  final bool enabled;
  final int ridesRequired;
  final bool repeats;

  factory LoyaltyConfig.fromMap(Map<String, dynamic>? data) {
    if (data == null) return const LoyaltyConfig();
    return LoyaltyConfig(
      enabled: data['enabled'] as bool? ?? false,
      ridesRequired: (data['ridesRequired'] as num?)?.toInt() ?? 10,
      repeats: data['repeats'] as bool? ?? true,
    );
  }

  Map<String, dynamic> toMap() => {
        'enabled': enabled,
        'ridesRequired': ridesRequired < 1 ? 1 : ridesRequired,
        'repeats': repeats,
      };

  String summary(bool isAr) {
    if (!enabled) {
      return isAr ? 'ولاء العملاء: معطّل' : 'Customer loyalty: off';
    }
    if (repeats) {
      return isAr
          ? 'مشوار مجاني كل $ridesRequired رحلات'
          : 'Free ride every $ridesRequired trips';
    }
    return isAr
        ? 'مشوار مجاني مرة عند $ridesRequired رحلات'
        : 'One free ride at $ridesRequired trips';
  }
}

class DriverMonthlyStats {
  const DriverMonthlyStats({
    required this.rideCount,
    required this.rank,
    required this.totalDrivers,
    required this.prizeAmountIqd,
    required this.monthKey,
  });

  final int rideCount;
  final int rank;
  final int totalDrivers;
  final int prizeAmountIqd;
  final String monthKey;
}

File: hilla_ride_ios/HelloTukTuk/Core/Models/PromoModels.swift
import FirebaseFirestore
import Foundation

struct PromoCodeConfig {
    let code: String
    let enabled: Bool
    let discountPercent: Int
    let maxDiscountIqd: Int
    let maxRides: Int

    static let free3Defaults = PromoCodeConfig(
        code: "FREE3",
        enabled: true,
        discountPercent: 50,
        maxDiscountIqd: 1000,
        maxRides: 2
    )

    init(code: String, enabled: Bool, discountPercent: Int, maxDiscountIqd: Int, maxRides: Int) {
        self.code = code
        self.enabled = enabled
        self.discountPercent = discountPercent
        self.maxDiscountIqd = maxDiscountIqd
        self.maxRides = maxRides
    }

    init?(data: [String: Any]?) {
        guard let data else {
            self = .free3Defaults
            return
        }
        code = data["code"] as? String ?? PromoCodeConfig.free3Defaults.code
        enabled = data["enabled"] as? Bool ?? true
        discountPercent = (data["discountPercent"] as? NSNumber)?.intValue ?? 50
        maxDiscountIqd = (data["maxDiscountIqd"] as? NSNumber)?.intValue ?? 1000
        maxRides = (data["maxRides"] as? NSNumber)?.intValue ?? 2
    }
}

struct PromoApplication {
    let baseFareIqd: Int
    let discountIqd: Int
    let finalFareIqd: Int
    let promoCode: String

    var hasDiscount: Bool { discountIqd > 0 && !promoCode.isEmpty }
}

struct MonthlyPrizeConfig: Equatable {
    let prizeAmountIqd: Int
    let monthKey: String
    let winnerDriverId: String
    let winnerPaid: Bool

    static let defaultPrizeIqd = 50_000

    static func currentMonthKey(from date: Date = Date()) -> String {
        let calendar = Calendar.current
        let year = calendar.component(.year, from: date)
        let month = calendar.component(.month, from: date)
        return String(format: "%04d-%02d", year, month)
    }

    init(
        prizeAmountIqd: Int = defaultPrizeIqd,
        monthKey: String = MonthlyPrizeConfig.currentMonthKey(),
        winnerDriverId: String = "",
        winnerPaid: Bool = false
    ) {
        self.prizeAmountIqd = prizeAmountIqd
        self.monthKey = monthKey
        self.winnerDriverId = winnerDriverId
        self.winnerPaid = winnerPaid
    }

    init(data: [String: Any]?) {
        guard let data else {
            self = MonthlyPrizeConfig()
            return
        }
        prizeAmountIqd = (data["prizeAmountIqd"] as? NSNumber)?.intValue ?? Self.defaultPrizeIqd
        monthKey = data["monthKey"] as? String ?? Self.currentMonthKey()
        winnerDriverId = data["winnerDriverId"] as? String ?? ""
        winnerPaid = data["winnerPaid"] as? Bool ?? false
    }
}

struct DriverMonthlyStats: Equatable {
    let rideCount: Int
    let rank: Int
    let totalDrivers: Int
    let prizeAmountIqd: Int
    let monthKey: String
}

final class MonthlyPrizeService {
    private let firestore = Firestore.firestore()

    func getPromoCode(_ code: String) async -> PromoCodeConfig {
        do {
            let doc = try await firestore.collection("config").document("promo_\(code)").getDocument()
            return PromoCodeConfig(data: doc.data()) ?? .free3Defaults
        } catch {
            return .free3Defaults
        }
    }

    func applyPromo(user: AppUser, config: PromoCodeConfig, baseFareIqd: Int) -> PromoApplication {
        guard baseFareIqd > 0,
              config.enabled,
              user.promoCode == config.code,
              user.promoRidesUsed < user.promoRidesLimit else {
            return PromoApplication(
                baseFareIqd: baseFareIqd,
                discountIqd: 0,
                finalFareIqd: baseFareIqd,
                promoCode: ""
            )
        }

        let rawDiscount = Int((Double(baseFareIqd) * Double(config.discountPercent) / 100.0).rounded())
        let discount = min(max(rawDiscount, 0), config.maxDiscountIqd)
        let finalFare = max(baseFareIqd - discount, 0)
        return PromoApplication(
            baseFareIqd: baseFareIqd,
            discountIqd: discount,
            finalFareIqd: finalFare,
            promoCode: config.code
        )
    }
}

struct MonthlyPrizeConfig: Equatable {
    let prizeAmountIqd: Int
    let monthKey: String
    let winnerDriverId: String
    let winnerPaid: Bool

    static let defaultPrizeIqd = 50_000

    static func currentMonthKey(from date: Date = Date()) -> String {
        let calendar = Calendar.current
        let year = calendar.component(.year, from: date)
        let month = calendar.component(.month, from: date)
        return String(format: "%04d-%02d", year, month)
    }

    init(
        prizeAmountIqd: Int = defaultPrizeIqd,
        monthKey: String = MonthlyPrizeConfig.currentMonthKey(),
        winnerDriverId: String = "",
        winnerPaid: Bool = false
    ) {
        self.prizeAmountIqd = prizeAmountIqd
        self.monthKey = monthKey
        self.winnerDriverId = winnerDriverId
        self.winnerPaid = winnerPaid
    }

    init(data: [String: Any]?) {
        guard let data else {
            self = MonthlyPrizeConfig()
            return
        }
        prizeAmountIqd = (data["prizeAmountIqd"] as? NSNumber)?.intValue ?? Self.defaultPrizeIqd
        monthKey = data["monthKey"] as? String ?? Self.currentMonthKey()
        winnerDriverId = data["winnerDriverId"] as? String ?? ""
        winnerPaid = data["winnerPaid"] as? Bool ?? false
    }
}

struct DriverMonthlyStats: Equatable {
    let rideCount: Int
    let rank: Int
    let totalDrivers: Int
    let prizeAmountIqd: Int
    let monthKey: String
}

final class MonthlyPrizeService {
    private let firestore = Firestore.firestore()

    func getConfig() async -> MonthlyPrizeConfig {
        do {
            let doc = try await firestore.collection("config").document("monthly_prize").getDocument()
            return MonthlyPrizeConfig(data: doc.data())
        } catch {
            return MonthlyPrizeConfig()
        }
    }

    func watchDriverStats(driverId: String) -> AsyncStream<DriverMonthlyStats> {
        AsyncStream { continuation in
            let configListener = firestore.collection("config").document("monthly_prize")
                .addSnapshotListener { _, _ in }
            let driversListener = firestore.collection("drivers")
                .addSnapshotListener { [weak self] snapshot, _ in
                    guard let self else { return }
                    Task {
                        let config = await self.getConfig()
                        let stats = self.buildStats(
                            driverId: driverId,
                            config: config,
                            docs: snapshot?.documents ?? []
                        )
                        continuation.yield(stats)
                    }
                }
            continuation.onTermination = { _ in
                configListener.remove()
                driversListener.remove()
            }
        }
    }

    func incrementDriverMonthlyRide(driverId: String) async {
        guard !driverId.isEmpty else { return }
        let monthKey = MonthlyPrizeConfig.currentMonthKey()
        let ref = firestore.collection("drivers").document(driverId)
        let snapshot = try? await ref.getDocument()
        guard let data = snapshot?.data() else { return }
        let currentKey = data["monthlyMonthKey"] as? String ?? ""
        if currentKey != monthKey {
            try? await ref.updateData([
                "monthlyMonthKey": monthKey,
                "monthlyRideCount": 1
            ])
        } else {
            try? await ref.updateData([
                 "monthlyRideCount": FieldValue.increment(Int64(1))
            ])
        }
    }

    private func buildStats(
        driverId: String,
        config: MonthlyPrizeConfig,
        docs: [QueryDocumentSnapshot]
    ) -> DriverMonthlyStats {
        var rows: [(id: String, name: String, count: Int)] = []
        for doc in docs {
            guard let driver = DriverProfile(documentID: doc.documentID, data: doc.data()),
                  driver.isApproved,
                  driver.monthlyMonthKey == config.monthKey,
                  driver.monthlyRideCount > 0 else { continue }
            rows.append((driver.uid, driver.name, driver.monthlyRideCount))
        }
        rows.sort {
            if $0.count != $1.count { return $0.count > $1.count }
            return $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending
        }
        let rank = (rows.firstIndex { $0.id == driverId }).map { $0 + 1 }
            ?? (rows.isEmpty ? 1 : rows.count + 1)
        let rideCount = rows.first(where: { $0.id == driverId })?.count ?? 0
        return DriverMonthlyStats(
            rideCount: rideCount,
            rank: rank,
            totalDrivers: rows.count,
            prizeAmountIqd: config.prizeAmountIqd,
            monthKey: config.monthKey
        )
    }
}
